File: pyRBDL/test/octave/mRBDL/Kinematics/CalcWholeBodyCoM.m
function Pcom = CalcWholeBodyCoM(model, q)
%CalcWholeBodyCoM - Calculate whole body's CoM position in world frame
%
% Syntax: Pcom = CalcWholeBodyCoM(q)
%
% Long description


    num = max(size(model.idcomplot,1), size(model.idcomplot,2));


    % Calcualte link's CoM in world frame
    CoM = [];
    Clink = zeros(3,1);
    for i=1:num
        Clink = CalcBodyToBaseCoordinates(model, q, model.idcomplot(i), model.CoM{i});
        CoM = [CoM, Clink]; 
    end

    % Calculate whole CoM in world frame    
    C = zeros(3,1);
    M = 0;
    for i=1:num
        C = C + CoM(:, i) * model.Mass{i};
        M = M + model.Mass{i};
    end
    Pcom = C/M;
end

File: pyRBDL/Kinematics/CalcWholeBodyCoM.py
import numpy as np
from pyRBDL.Kinematics.CalcBodyToBaseCoordinates import CalcBodyToBaseCoordinates


def CalcWholeBodyCoM(model: dict, q: np.ndarray)-> np.ndarray:
    """CalcWholeBodyCoM - Calculate whole body's CoM position in world frame

    Args:
        model (dict): dictionary of model specification
        q (np.ndarray): an array of joint position

    Returns:
        np.ndarray: float (3, 3)
    """  
    q = q.flatten()
    idcomplot = model["idcomplot"].flatten().astype(int)
    try:
        CoM = np.squeeze(model["CoM"], axis=0)
        Mass = np.squeeze(model["Mass"], axis=0)
    except:
        CoM = model["CoM"]
        Mass = model["Mass"]
    
    
    num = np.max(idcomplot.shape)
    CoM_list = []
    Clink = np.zeros((3,1))
    for i in range(num):
        Clink = CalcBodyToBaseCoordinates(model, q, idcomplot[i], CoM[i])
        CoM_list.append(Clink)
    
    C = np.zeros((3, 1))
    M = 0

    for i in range(num):
        C = C + np.multiply(CoM_list[i], Mass[i])
        M = M + Mass[i]

    Pcom = np.asfarray(np.divide(C, M))
    
    return Pcom

